axlearn/cloud/gcp/tpu.py

```diff
@@ -567,9 +569,21 @@ def _tpu_body(
             "enable_ici_resiliency": metadata.pop("enable_ici_resiliency"),
         }
 
+    vm_tier_labels = {}
+    # Create labels for vm tier that can be used to group tpu metrics.
+    # TPUs are created via QRM. In QRM, vm tier can be one of guaranteed, spot, or, bestEffort.
+    # guaranteeded is mapped to label value reserved. bestEffort is implemented to create
+    # spot instance. Hence, both spot and bestEffort are mapped to lable value spot.
+    # See [`_qrm_body()`](#L651) for details on the tiers.
+    if reserved or gcp_settings("reserved_tpu", default=False):
+        vm_tier_labels["vmtier"] = "reserved"
+    else:
+        vm_tier_labels["vmtier"] = "spot"
+
```

Comment:
Instead of special-casing the vmtier label here, how about letting users supply a dict of `labels` in the TPU creation, similar to `metadata`? That way, the tier logic can be consolidated in the bastion code.

Reply:
@markblee  

> how about letting users supply a dict of labels in the TPU creation, similar to metadata

Are you talking about adding labels as a param in [create_queued_tpu()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L293) function  and pass it to the [_tpu_body()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L339)? 

> That way, the tier logic can be consolidated in the bastion code

Can you point me to the bastion code?

Reply:
> Are you talking about adding labels as a param in [create_queued_tpu()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L293) function and pass it to the [_tpu_body()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L339)?

That's right.

> Can you point me to the bastion code?

This part might be of interest: https://github.com/apple/axlearn/blob/406c3840480422755deb3f9e27806f368f2b7d48/axlearn/cloud/gcp/jobs/tpu_runner.py#L303-L312

Reply:
@markblee please take a look at the [PR 544](https://github.com/apple/axlearn/pull/544),  which has changes that you suggested.